container-search/src/main/java/com/yahoo/prelude/fastsearch/FastHit.java

```diff
@@ -205,4 +354,402 @@ public static String asHexString(GlobalId gid) {
+
+            protected void advanceNext() {
+                while (fieldIterator.hasNext()) {
+                    Map.Entry<String, Inspector> nextEntry = fieldIterator.next();
+                    String fieldName = nextEntry.getKey();
+                    next = toValue(nextEntry);
+                    if ( next != null &&
+                         ! hit.hasField(fieldName) &&
+                         ! isRemoved(fieldName) &&
+                         ! isAlreadyReturned(fieldName))
+                        return;
+                }
+                next = null;
+            }
+
+            private boolean isRemoved(String fieldName) {
+                return hit.removedFields != null && hit.removedFields.contains(fieldName);
+            }
+
+            private boolean isAlreadyReturned(String fieldName) {
+                for (int i = 0; i < index; i++) {
+                    if (hit.summaries.get(i).type.fieldNames().contains(fieldName))
+                        return true;
+                }
+                return false;
+            }
+
+        }
+
+        private static class SummaryFieldDataIterator extends SummaryDataIterator<Map.Entry<String, Object>> {
+
+            private final DocsumDefinition type;
+
+            SummaryFieldDataIterator(FastHit hit,
+                                     DocsumDefinition type,
+                                     Iterator<Map.Entry<String, Inspector>> fieldIterator,
+                                     int index) {
+                super(hit, fieldIterator, index);
+                this.type = type;
+                advanceNext();
+            }
+
+            @Override
+            protected Map.Entry<String, Object> toValue(Map.Entry<String, Inspector> field) {
+                DocsumField fieldType = type.getField(field.getKey());
+                if (fieldType == null) return null; // type and content mismatch: May happen during reconfig
+                return new SummaryFieldEntry(field.getKey(), fieldType.convert(field.getValue()));
+            }
+
+            private static final class SummaryFieldEntry implements Map.Entry<String, Object> {
+
+                private final String key;
+                private final Object value;
+
+                public SummaryFieldEntry(String key, Object value) {
+                    this.key = key;
+                    this.value = value;
+                }
+
+                @Override
+                public String getKey() { return key; }
+
+                @Override
+                public Object getValue() { return value; }
+
+                @Override
+                public Object setValue(Object value) { throw new UnsupportedOperationException(); }
+
+            }
+
+        }
+
+        /** A wrapper of a field iterator which converts to the expected field types. Read only. */
+        private static class SummaryFieldNameDataIterator extends SummaryDataIterator<String> {
```

Comment:
A short comment on the purpose of the class would be good, maybe something like:
Iterate over the field names of a single summary data object.
Also consider changing name to 'SummaryDataFieldNameIterator'.